### PR TITLE
Transformer Encoder

### PR DESCRIPTION
**Pull request by JesusPinedaC**
This pull request introduces the `TransformerEncoderLayer` component as outlined in "[Attention Is All You Need](https://arxiv.org/abs/1706.03762)". Additionally, the pull request brings in a new processing block called `LayerSkipNormalization` and the DeeplayModule `MultiheadSelfAttention`.   

The `MultiheadSelfAttention` module is designed to handle both batched and unbatched sequences. When processing unbatched inputs, the module incorporates a mechanism where `batch_indices` are accepted. These indices help condition the attention weights of each item, ensuring that they exclusively attend to other items within the same batch.

### CODE REVIEW

File: deeplay/components/transformer/enc.py
Comment (Henrik-KM) [inline deeplay/components/transformer/enc.py:1]: May be more deeplay-like to use dl MLP here rather than torch modules in a sequential?

File: deeplay/components/transformer/enc.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+class Add(DeeplayModule):
+    def forward(self, *x):
+        """
+        Add the input tensors element-wise.
+        """
+        x = itertools.chain(*x) if isinstance(x[0], tuple) else x
+        return reduce(lambda a, b: torch.add(a, b), x)
+
+
+class TransformerEncoderLayer(DeeplayModule):
```
Comment (Henrik-KM) [inline deeplay/components/transformer/enc.py:32]: Should contain some documentation in comments, similarly to the the other components

File: deeplay/components/transformer/enc.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                            if f_in != f_out
+                            else Layer(nn.Identity),
+                        ),
+                        skip=Add(),
+                        normalization=Layer(nn.LayerNorm, f_out),
+                    ),
+                    # Second sub-block is feed forward followed by skip connection
+                    # and normalization
+                    feed_forward=LayerSkipNormalization(
+                        layer=Sequential(
```
Comment (Henrik-KM) [inline deeplay/components/transformer/enc.py]: May be more deeplay-like to use dl MLP here rather than torch modules in a sequential?

File: deeplay/components/transformer/satt.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
+from deeplay import DeeplayModule, Layer
+
+import torch
+import torch.nn as nn
+
+
+class MultiheadSelfAttention(DeeplayModule):
```
Comment (Henrik-KM) [inline deeplay/components/transformer/satt.py:7]: Should also contain some form of documentation, e.g. what is now in the forward-pass

**Review by Henrik-KM — APPROVED** [review]

**Review by Henrik-KM — APPROVED** [review]

File: deeplay/components/transformer/enc.py (hunk 2 above)
Comment (JesusPinedaC) [reply deeplay/components/transformer/enc.py]: Yeah, it is a good idea. 
My main concern with this is that it could increase the intricacy of the model (already quite complex), especially when printed, by adding several `normalization=nn.Identity` throughout the model. @Henrik-KM

File: deeplay/components/transformer/satt.py (hunk 3 above)
Comment (JesusPinedaC) [reply deeplay/components/transformer/satt.py:7]: Done!

File: deeplay/components/transformer/enc.py (hunk 1 above)
Comment (JesusPinedaC) [reply deeplay/components/transformer/enc.py:32]: Done!

File: deeplay/components/transformer/enc.py (hunk 2 above)
Comment (BenjaminMidtvedt) [reply deeplay/components/transformer/enc.py]: I agree with henrik here. We can maybe think of a way to reduce visual clutter to not display identity layers as prominently

Comment (JesusPinedaC) [reply deeplay/components/transformer/enc.py]: Modified by [40bac2c](https://github.com/DeepTrackAI/deeplay/pull/69/commits/40bac2cdf4e7ddfc13914a682534ee124df3aae9)

**Review by BenjaminMidtvedt — APPROVED** [review]
